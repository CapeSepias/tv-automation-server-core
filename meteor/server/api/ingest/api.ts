import * as _ from 'underscore'
import { PeripheralDeviceAPI } from '../../../lib/api/peripheralDevice'
import { IngestRundown, IngestSegment, IngestPart } from 'tv-automation-sofie-blueprints-integration'
import { Methods, setMeteorMethods } from '../../methods'

import { RundownInput } from './rundownInput'
import { PeripheralDeviceId } from '../../../lib/collections/PeripheralDevices'

let methods: Methods = {}

methods[PeripheralDeviceAPI.methods.dataRundownList] = (deviceId: PeripheralDeviceId, deviceToken: string) => {
	return RundownInput.dataRundownList(this, deviceId, deviceToken)
}
methods[PeripheralDeviceAPI.methods.dataRundownGet] = (deviceId: PeripheralDeviceId, deviceToken: string, rundownExternalId: string) => {
	return RundownInput.dataRundownGet(this, deviceId, deviceToken, rundownExternalId)
}
methods[PeripheralDeviceAPI.methods.dataRundownDelete] = (deviceId: PeripheralDeviceId, deviceToken: string, rundownExternalId: string) => {
	return RundownInput.dataRundownDelete(this, deviceId, deviceToken, rundownExternalId)
}
methods[PeripheralDeviceAPI.methods.dataRundownCreate] = (deviceId: PeripheralDeviceId, deviceToken: string, ingestRundown: IngestRundown) => {
	return RundownInput.dataRundownCreate(this, deviceId, deviceToken, ingestRundown)
}
methods[PeripheralDeviceAPI.methods.dataRundownUpdate] = (deviceId: PeripheralDeviceId, deviceToken: string, ingestRundown: IngestRundown) => {
	return RundownInput.dataRundownUpdate(this, deviceId, deviceToken, ingestRundown)
}
methods[PeripheralDeviceAPI.methods.dataSegmentDelete] = (deviceId: PeripheralDeviceId, deviceToken: string, rundownExternalId: string, segmentExternalId: string) => {
	return RundownInput.dataSegmentDelete(this, deviceId, deviceToken, rundownExternalId, segmentExternalId)
}
methods[PeripheralDeviceAPI.methods.dataSegmentCreate] = (deviceId: PeripheralDeviceId, deviceToken: string, rundownExternalId: string, ingestSegment: IngestSegment) => {
	return RundownInput.dataSegmentCreate(this, deviceId, deviceToken, rundownExternalId, ingestSegment)
}
methods[PeripheralDeviceAPI.methods.dataSegmentUpdate] = (deviceId: PeripheralDeviceId, deviceToken: string, rundownExternalId: string, ingestSegment: IngestSegment) => {
	return RundownInput.dataSegmentUpdate(this, deviceId, deviceToken, rundownExternalId, ingestSegment)
}
methods[PeripheralDeviceAPI.methods.dataPartDelete] = (deviceId: PeripheralDeviceId, deviceToken: string, rundownExternalId: string, segmentExternalId: string, partExternalId: string) => {
	return RundownInput.dataPartDelete(this, deviceId, deviceToken, rundownExternalId, segmentExternalId, partExternalId)
}
methods[PeripheralDeviceAPI.methods.dataPartCreate] = (deviceId: PeripheralDeviceId, deviceToken: string, rundownExternalId: string, segmentExternalId: string, ingestPart: IngestPart) => {
	return RundownInput.dataPartCreate(this, deviceId, deviceToken, rundownExternalId, segmentExternalId, ingestPart)
}
methods[PeripheralDeviceAPI.methods.dataPartUpdate] = (deviceId: PeripheralDeviceId, deviceToken: string, rundownExternalId: string, segmentExternalId: string, ingestPart: IngestPart) => {
	return RundownInput.dataPartUpdate(this, deviceId, deviceToken, rundownExternalId, segmentExternalId, ingestPart)
}

setMeteorMethods(methods)
