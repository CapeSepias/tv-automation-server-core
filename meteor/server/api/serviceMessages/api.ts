import { IncomingMessage, ServerResponse } from 'http'
import { Picker } from 'meteor/meteorhacks:picker'
import { CoreSystem } from '../../../lib/collections/CoreSystem'
import { logger } from '../../logging';

Picker.route('/serviceMessages', (params, req, res, next) => {
	const { method } = req

	if (method === 'POST') {
		return postHandler(params, req, res, next)
	}
	if (method == 'GET') {
		return getHandler(params, req, res, next)
	}

	res.statusCode = 405
	res.end()
})

Picker.route('/serviceMessages/:id', (params, req, res, next) => {
	const { method } = req

	if (method === 'DELETE') {
		return deleteHandler(params, req, res, next)
	}
	if (method == 'GET') {
		return getMessageHandler(params, req, res, next)
	}

	res.statusCode = 405
	res.end()
})

/**
 * Create new or update existing message
 */
function postHandler (
	params,
	req: IncomingMessage,
	res: ServerResponse,
	next: () => void
) {
	// validate !? => 400
	// exists? update exisiting => 200/204
	// create => 201
	// fuckup? => 500
}

/**
 * List all current messages stored on this instance
 */
function getHandler (
	params,
	req: IncomingMessage,
	res: ServerResponse,
	next: () => void
) {
	const coreSystem = CoreSystem.findOne()
	if (!coreSystem || !coreSystem.serviceMessages) {
		return
	}

	const { serviceMessages } = coreSystem
	logger.info(`serviceMessages: ${typeof serviceMessages}`, serviceMessages)
	const valuesArray = Array.from(Object.entries(serviceMessages))
	logger.info(`valuesArray: ${valuesArray.length}, ${typeof valuesArray}`)
	res.end(JSON.stringify(valuesArray))
}

/**
 * Delete a message
 */
function deleteHandler (
	params,
	req: IncomingMessage,
	res: ServerResponse,
	next: () => void
) {
	// exists ? delete => 200/204 : 404
}

/**
 * Retrieves a single message based on a given id
 */
function getMessageHandler (
	params,
	req: IncomingMessage,
	res: ServerResponse,
	next: () => void
) {
	// exists ? => json object, 200 : => 404
}
