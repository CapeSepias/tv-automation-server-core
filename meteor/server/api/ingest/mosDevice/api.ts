import * as MOS from 'mos-connection'
import { setMeteorMethods, Methods } from '../../../methods'
import { PeripheralDeviceAPI } from '../../../../lib/api/peripheralDevice'
import { MosIntegration } from './mosIntegration'
import { PeripheralDeviceId } from '../../../../lib/collections/PeripheralDevices'

let methods: Methods = {}
methods[PeripheralDeviceAPI.methods.mosRoCreate] = (deviceId: PeripheralDeviceId, deviceToken: string, mosRunningOrder: MOS.IMOSRunningOrder) => {
	return MosIntegration.mosRoCreate(deviceId, deviceToken, mosRunningOrder)
}
methods[PeripheralDeviceAPI.methods.mosRoReplace] = (deviceId: PeripheralDeviceId, deviceToken: string, mosRunningOrder: MOS.IMOSRunningOrder) => {
	return MosIntegration.mosRoReplace(deviceId, deviceToken, mosRunningOrder)
}
methods[PeripheralDeviceAPI.methods.mosRoDelete] = (deviceId: PeripheralDeviceId, deviceToken: string, mosRunningOrderId: MOS.MosString128, force?: boolean) => {
	return MosIntegration.mosRoDelete(deviceId, deviceToken, mosRunningOrderId, force)
}
methods[PeripheralDeviceAPI.methods.mosRoMetadata] = (deviceId: PeripheralDeviceId, deviceToken: string, metadata: MOS.IMOSRunningOrderBase) => {
	return MosIntegration.mosRoMetadata(deviceId, deviceToken, metadata)
}
methods[PeripheralDeviceAPI.methods.mosRoStatus] = (deviceId: PeripheralDeviceId, deviceToken: string, status: MOS.IMOSRunningOrderStatus) => {
	return MosIntegration.mosRoStatus(deviceId, deviceToken, status)
}
methods[PeripheralDeviceAPI.methods.mosRoStoryStatus] = (deviceId: PeripheralDeviceId, deviceToken: string, status: MOS.IMOSStoryStatus) => {
	return MosIntegration.mosRoStoryStatus(deviceId, deviceToken, status)
}
methods[PeripheralDeviceAPI.methods.mosRoItemStatus] = (deviceId: PeripheralDeviceId, deviceToken: string, status: MOS.IMOSItemStatus) => {
	return MosIntegration.mosRoItemStatus(deviceId, deviceToken, status)
}
methods[PeripheralDeviceAPI.methods.mosRoStoryInsert] = (deviceId: PeripheralDeviceId, deviceToken: string, Action: MOS.IMOSStoryAction, Stories: Array<MOS.IMOSROStory>) => {
	return MosIntegration.mosRoStoryInsert(deviceId, deviceToken, Action, Stories)
}
methods[PeripheralDeviceAPI.methods.mosRoItemInsert] = (deviceId: PeripheralDeviceId, deviceToken: string, Action: MOS.IMOSItemAction, Items: Array<MOS.IMOSItem>) => {
	return MosIntegration.mosRoItemInsert(deviceId, deviceToken, Action, Items)
}
methods[PeripheralDeviceAPI.methods.mosRoStoryReplace] = (deviceId: PeripheralDeviceId, deviceToken: string, Action: MOS.IMOSStoryAction, Stories: Array<MOS.IMOSROStory>) => {
	return MosIntegration.mosRoStoryReplace(deviceId, deviceToken, Action, Stories)
}
methods[PeripheralDeviceAPI.methods.mosRoItemReplace] = (deviceId: PeripheralDeviceId, deviceToken: string, Action: MOS.IMOSItemAction, Items: Array<MOS.IMOSItem>) => {
	return MosIntegration.mosRoItemReplace(deviceId, deviceToken, Action, Items)
}
methods[PeripheralDeviceAPI.methods.mosRoStoryMove] = (deviceId: PeripheralDeviceId, deviceToken: string, Action: MOS.IMOSStoryAction, Stories: Array<MOS.MosString128>) => {
	return MosIntegration.mosRoStoryMove(deviceId, deviceToken, Action, Stories)
}
methods[PeripheralDeviceAPI.methods.mosRoItemMove] = (deviceId: PeripheralDeviceId, deviceToken: string, Action: MOS.IMOSItemAction, Items: Array<MOS.MosString128>) => {
	return MosIntegration.mosRoItemMove(deviceId, deviceToken, Action, Items)
}
methods[PeripheralDeviceAPI.methods.mosRoStoryDelete] = (deviceId: PeripheralDeviceId, deviceToken: string, Action: MOS.IMOSROAction, Stories: Array<MOS.MosString128>) => {
	return MosIntegration.mosRoStoryDelete(deviceId, deviceToken, Action, Stories)
}
methods[PeripheralDeviceAPI.methods.mosRoItemDelete] = (deviceId: PeripheralDeviceId, deviceToken: string, Action: MOS.IMOSStoryAction, Items: Array<MOS.MosString128>) => {
	return MosIntegration.mosRoItemDelete(deviceId, deviceToken, Action, Items)
}
methods[PeripheralDeviceAPI.methods.mosRoStorySwap] = (deviceId: PeripheralDeviceId, deviceToken: string, Action: MOS.IMOSROAction, StoryID0: MOS.MosString128, StoryID1: MOS.MosString128) => {
	return MosIntegration.mosRoStorySwap(deviceId, deviceToken, Action, StoryID0, StoryID1)
}
methods[PeripheralDeviceAPI.methods.mosRoItemSwap] = (deviceId: PeripheralDeviceId, deviceToken: string, Action: MOS.IMOSStoryAction, ItemID0: MOS.MosString128, ItemID1: MOS.MosString128) => {
	return MosIntegration.mosRoItemSwap(deviceId, deviceToken, Action, ItemID0, ItemID1)
}
methods[PeripheralDeviceAPI.methods.mosRoReadyToAir] = (deviceId: PeripheralDeviceId, deviceToken: string, Action: MOS.IMOSROReadyToAir) => {
	return MosIntegration.mosRoReadyToAir(deviceId, deviceToken, Action)
}
methods[PeripheralDeviceAPI.methods.mosRoFullStory] = (deviceId: PeripheralDeviceId, deviceToken: string, story: MOS.IMOSROFullStory) => {
	return MosIntegration.mosRoFullStory(deviceId, deviceToken, story)
}

// Apply methods:
setMeteorMethods(methods)
