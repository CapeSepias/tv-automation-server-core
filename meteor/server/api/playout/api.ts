import { Methods, setMeteorMethods } from '../../methods'
import { PlayoutAPI } from '../../../lib/api/playout'
import { ServerPlayoutAPI } from './playout'
import { getCurrentTime } from '../../../lib/lib'
import { logger } from '../../logging'
import { RundownPlaylistId } from '../../../lib/collections/RundownPlaylists'
import { PartInstanceId } from '../../../lib/collections/PartInstances'
import { PartId } from '../../../lib/collections/Parts'
import { PieceId } from '../../../lib/collections/Pieces'
import { StudioId } from '../../../lib/collections/Studios'
import { PieceInstanceId } from '../../../lib/collections/PieceInstances'

let methods: Methods = {}
methods[PlayoutAPI.methods.rundownPrepareForBroadcast] = (playlistId: RundownPlaylistId) => {
	return ServerPlayoutAPI.prepareRundownPlaylistForBroadcast(playlistId)
}
methods[PlayoutAPI.methods.rundownResetRundown] = (playlistId: RundownPlaylistId) => {
	return ServerPlayoutAPI.resetRundownPlaylist(playlistId)
}
methods[PlayoutAPI.methods.rundownResetAndActivate] = (playlistId: RundownPlaylistId, rehearsal?: boolean) => {
	return ServerPlayoutAPI.resetAndActivateRundownPlaylist(playlistId, rehearsal)
}
methods[PlayoutAPI.methods.rundownActivate] = (playlistId: RundownPlaylistId, rehearsal: boolean) => {
	return ServerPlayoutAPI.activateRundownPlaylist(playlistId, rehearsal)
}
methods[PlayoutAPI.methods.rundownDeactivate] = (playlistId: RundownPlaylistId) => {
	return ServerPlayoutAPI.deactivateRundownPlaylist(playlistId)
}
methods[PlayoutAPI.methods.reloadRundownPlaylistData] = (playlistId: RundownPlaylistId) => {
	return ServerPlayoutAPI.reloadRundownPlaylistData(playlistId)
}
methods[PlayoutAPI.methods.pieceTakeNow] = (playlistId: RundownPlaylistId, partInstanceId: PartInstanceId, pieceInstanceIdOrPieceIdToCopy: PieceInstanceId | PieceId) => {
	return ServerPlayoutAPI.pieceTakeNow(playlistId, partInstanceId, pieceInstanceIdOrPieceIdToCopy)
}
methods[PlayoutAPI.methods.rundownTake] = (playlistId: RundownPlaylistId) => {
	return ServerPlayoutAPI.takeNextPart(playlistId)
}
methods[PlayoutAPI.methods.rundownTogglePartArgument] = (playlistId: RundownPlaylistId, partInstanceId: PartInstanceId, property: string, value: string) => {
	return ServerPlayoutAPI.rundownTogglePartArgument(playlistId, partInstanceId, property, value)
}
methods[PlayoutAPI.methods.rundownSetNext] = (playlistId: RundownPlaylistId, partId: PartId, timeOffset?: number | undefined) => {
	return ServerPlayoutAPI.setNextPart(playlistId, partId, true, timeOffset)
}
methods[PlayoutAPI.methods.rundownMoveNext] = (playlistId: RundownPlaylistId, horisontalDelta: number, verticalDelta: number) => {
	return ServerPlayoutAPI.moveNextPart(playlistId, horisontalDelta, verticalDelta, true)
}
methods[PlayoutAPI.methods.rundownActivateHold] = (playlistId: RundownPlaylistId) => {
	return ServerPlayoutAPI.activateHold(playlistId)
}
methods[PlayoutAPI.methods.rundownDisableNextPiece] = (rundownPlaylistId: RundownPlaylistId, undo?: boolean) => {
	return ServerPlayoutAPI.disableNextPiece(rundownPlaylistId, undo)
}
// methods[PlayoutAPI.methods.partPlaybackStartedCallback] = (rundownId: RundownId, partId: PartId, startedPlayback: number) => {
// 	return ServerPlayoutAPI.onPartPlaybackStarted(rundownId, partId, startedPlayback)
// }
// methods[PlayoutAPI.methods.piecePlaybackStartedCallback] = (rundownId: RundownId, pieceId: PieceId, startedPlayback: number) => {
// 	return ServerPlayoutAPI.onPiecePlaybackStarted(rundownId, pieceId, startedPlayback)
// }
methods[PlayoutAPI.methods.segmentAdLibPieceStart] = (rundownPlaylistId: RundownPlaylistId, partInstanceId: PartInstanceId, pieceId: PieceId, queue: boolean) => {
	return ServerPlayoutAPI.segmentAdLibPieceStart(rundownPlaylistId, partInstanceId, pieceId, queue)
}
methods[PlayoutAPI.methods.rundownBaselineAdLibPieceStart] = (rundownPlaylistId: RundownPlaylistId, partInstanceId: PartInstanceId, pieceId: PieceId, queue: boolean) => {
	return ServerPlayoutAPI.rundownBaselineAdLibPieceStart(rundownPlaylistId, partInstanceId, pieceId, queue)
}
methods[PlayoutAPI.methods.sourceLayerOnPartStop] = (rundownPlaylistId: RundownPlaylistId, partInstanceId: PartInstanceId, sourceLayerId: string) => {
	return ServerPlayoutAPI.sourceLayerOnPartStop(rundownPlaylistId, partInstanceId, sourceLayerId)
}
// methods[PlayoutAPI.methods.timelineTriggerTimeUpdateCallback] = (timelineObjId: string, time: number) => {
// 	return ServerPlayoutAPI.timelineTriggerTimeUpdateCallback(timelineObjId, time)
// }
methods[PlayoutAPI.methods.sourceLayerStickyPieceStart] = (playlistId: RundownPlaylistId, sourceLayerId: string) => {
	return ServerPlayoutAPI.sourceLayerStickyPieceStart(playlistId, sourceLayerId)
}
methods[PlayoutAPI.methods.updateStudioBaseline] = (studioId: StudioId) => {
	return ServerPlayoutAPI.updateStudioBaseline(studioId)
}
methods[PlayoutAPI.methods.shouldUpdateStudioBaseline] = (studioId: StudioId) => {
	return ServerPlayoutAPI.shouldUpdateStudioBaseline(studioId)
}
// Apply methods:
setMeteorMethods(methods)

// Temporary methods
setMeteorMethods({
	'debug__printTime': () => {
		let now = getCurrentTime()
		logger.debug(new Date(now))
		return now
	},
})
