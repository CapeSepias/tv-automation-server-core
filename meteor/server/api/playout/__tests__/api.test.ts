import { Meteor } from 'meteor/meteor'
import '../../../../__mocks__/_extendJest'
import { testInFiber } from '../../../../__mocks__/helpers/jest'
import { setupDefaultStudioEnvironment, DefaultEnvironment, setupDefaultRundownPlaylist } from '../../../../__mocks__/helpers/database'
import { Rundowns, Rundown } from '../../../../lib/collections/Rundowns'
import '../api'
import { RundownPlaylists } from '../../../../lib/collections/RundownPlaylists'

namespace PlayoutAPI { // Using our own method definition, to catch external API changes
	export enum methods {

		'rundownPrepareForBroadcast' 		= 'playout.rundownPrepareForBroadcast',
		'rundownResetRundown' 				= 'playout.rundownResetRundownt',
		'rundownResetAndActivate' 			= 'playout.rundownResetAndActivate',
		'rundownActivate' 					= 'playout.rundownActivate',
		'rundownDeactivate' 				= 'playout.rundownDeactivate',
		'reloadData' 						= 'playout.reloadData',

		'updateStudioBaseline'				= 'playout.updateStudioBaseline',
		'shouldUpdateStudioBaseline'		= 'playout.shouldUpdateStudioBaseline',

		'rundownTake'						= 'playout.rundownTake',
		'rundownSetNext'					= 'playout.rundownSetNext',
		'rundownMoveNext'					= 'playout.rundownMoveNext',
		'rundownActivateHold'				= 'playout.rundownActivateHold',
		'rundownDisableNextPiece'			= 'playout.rundownDisableNextPiece',
		'rundownTogglePartArgument'			= 'playout.rundownTogglePartArgument',
		// 'partPlaybackStartedCallback'		= 'playout.partPlaybackStartedCallback',
		// 'piecePlaybackStartedCallback'		= 'playout.piecePlaybackStartedCallback',
		'pieceTakeNow'						= 'playout.pieceTakeNow',
		'segmentAdLibPieceStart'			= 'playout.segmentAdLibPieceStart',
		'rundownBaselineAdLibPieceStart'	= 'playout.rundownBaselineAdLibPieceStart',
		'segmentAdLibPieceStop'				= 'playout.segmentAdLibPieceStop',
		'sourceLayerOnPartStop'				= 'playout.sourceLayerOnPartStop',
		'sourceLayerStickyPieceStart'		= 'playout.sourceLayerStickyPieceStart'
	}
}

describe('Playout API', () => {
	let env: DefaultEnvironment
	beforeEach(() => {
		env = setupDefaultStudioEnvironment()
	})
	testInFiber('Basic rundown control', () => {
		const {
			rundownId: rundownId0,
			playlistId: playlistId0
		} = setupDefaultRundownPlaylist(env)
		const {
			rundownId: rundownId1,
			playlistId: playlistId1
		} = setupDefaultRundownPlaylist(env)
		expect(rundownId0).toBeTruthy()
		expect(rundownId1).toBeTruthy()
		expect(playlistId0).toBeTruthy()
		expect(playlistId1).toBeTruthy()


		const getRundown0 = () => {
			return Rundowns.findOne(rundownId0) as Rundown
		}
		const getRundown1 = () => {
			return Rundowns.findOne(rundownId1) as Rundown
		}
		const getPlaylist0 = () => {
			return RundownPlaylists.findOne(playlistId0)
		}
		const getPlaylist1 = () => {
			return RundownPlaylists.findOne(playlistId1)
		}

		expect(getRundown0()).toBeTruthy()
		expect(getRundown1()).toBeTruthy()
		expect(getRundown0()._id).not.toEqual(getRundown1()._id)
		expect(getPlaylist0()).toBeTruthy()
		expect(getPlaylist1()).toBeTruthy()

		const parts = getRundown0().getParts()

		expect(getPlaylist0()).toMatchObject({
			active: false,
			rehearsal: false
		})

		// Prepare and activate in rehersal:
		Meteor.call(PlayoutAPI.methods.rundownPrepareForBroadcast, playlistId0)
		expect(getPlaylist0()).toMatchObject({
			active: true,
			rehearsal: true,
			currentPartId: null,
			nextPartId: parts[0]._id,
		})
		// Activate a second rundown (this should throw an error)
		expect(() => {
			Meteor.call(PlayoutAPI.methods.rundownActivate, playlistId1)
		}).toThrowError(/only one rundown/i)


		// Take the first Part:
		Meteor.call(PlayoutAPI.methods.rundownTake, playlistId0)
		expect(getPlaylist0()).toMatchObject({
			currentPartId: parts[0]._id,
			nextPartId: parts[1]._id,
		})

		// Take the second Part:
		Meteor.call(PlayoutAPI.methods.rundownTake, playlistId0)
		expect(getPlaylist0()).toMatchObject({
			currentPartId: parts[1]._id,
			nextPartId: parts[2]._id,
		})
		// Reset rundown:
		Meteor.call(PlayoutAPI.methods.rundownResetRundown, playlistId0)
		expect(getPlaylist0()).toMatchObject({
			currentPartId: null,
			nextPartId: parts[0]._id,
		})

		// Set Part as next:
		Meteor.call(PlayoutAPI.methods.rundownSetNext, playlistId0, parts[parts.length - 2]._id)
		expect(getPlaylist0()).toMatchObject({
			currentPartId: null,
			nextPartId: parts[parts.length - 2]._id,
		})

		// Take the Nexted Part:
		Meteor.call(PlayoutAPI.methods.rundownTake, playlistId0)
		expect(getPlaylist0()).toMatchObject({
			currentPartId: parts[parts.length - 2]._id,
			nextPartId: parts[parts.length - 1]._id,
		})

		// Take the last Part:
		Meteor.call(PlayoutAPI.methods.rundownTake, playlistId0)
		expect(getPlaylist0()).toMatchObject({
			currentPartId: parts[parts.length - 1]._id,
			nextPartId: null
		})

		// Move the next-point backwards:
		Meteor.call(PlayoutAPI.methods.rundownMoveNext, playlistId0, -1, 0)
		expect(getPlaylist0()).toMatchObject({
			currentPartId: parts[parts.length - 1]._id,
			nextPartId: parts[parts.length - 2]._id
		})
		// Move the next-point backwards:
		Meteor.call(PlayoutAPI.methods.rundownMoveNext, playlistId0, -1, 0)
		expect(getPlaylist0()).toMatchObject({
			currentPartId: parts[parts.length - 1]._id,
			nextPartId: parts[parts.length - 3]._id
		})

		// Take the nexted Part:
		Meteor.call(PlayoutAPI.methods.rundownTake, playlistId0)
		expect(getPlaylist0()).toMatchObject({
			currentPartId: parts[parts.length - 3]._id,
			nextPartId: parts[parts.length - 2]._id,
		})

		// Deactivate rundown:
		Meteor.call(PlayoutAPI.methods.rundownDeactivate, playlistId0)
		expect(getPlaylist0()).toMatchObject({
			active: false,
			currentPartId: null,
			nextPartId: null
		})
	})
})
