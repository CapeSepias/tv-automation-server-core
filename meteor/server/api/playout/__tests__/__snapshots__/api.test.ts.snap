// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playout API Global and Part Ad-Libs 1`] = `
Array [
  Object {
    "_id": "randomId10002_part0_0_randomId10006_randomId10002_piece000",
    "partInstanceId": "randomId10002_part0_0_randomId10006",
    "piece": Object {
      "_id": "randomId10002_piece000",
      "enable": Object {
        "start": 0,
      },
      "externalId": "MOCK_PIECE_000",
      "name": "Piece 000",
      "outputLayerId": "pgm",
      "partId": "randomId10002_part0_0",
      "rundownId": "randomId10002",
      "sourceLayerId": "cam0",
      "status": 0,
    },
    "rundownId": "randomId10002",
  },
  Object {
    "_id": "randomId10002_part0_0_randomId10006_randomId10002_piece001",
    "partInstanceId": "randomId10002_part0_0_randomId10006",
    "piece": Object {
      "_id": "randomId10002_piece001",
      "enable": Object {
        "start": 0,
      },
      "externalId": "MOCK_PIECE_001",
      "name": "Piece 001",
      "outputLayerId": "pgm",
      "partId": "randomId10002_part0_0",
      "rundownId": "randomId10002",
      "sourceLayerId": "vt0",
      "status": 0,
    },
    "rundownId": "randomId10002",
  },
  Object {
    "_id": "randomId10002_part0_0_randomId10006_randomId10002_piece010",
    "partInstanceId": "randomId10002_part0_0_randomId10006",
    "piece": Object {
      "_id": "randomId10002_piece010",
      "enable": Object {
        "start": 0,
      },
      "externalId": "MOCK_PIECE_010",
      "name": "Piece 010",
      "outputLayerId": "pgm",
      "partId": "randomId10002_part0_0",
      "rundownId": "randomId10002",
      "sourceLayerId": "cam0",
      "status": 0,
    },
    "rundownId": "randomId10002",
  },
  Object {
    "_id": "randomId10002_part0_0_randomId10006_randomId10011",
    "partInstanceId": "randomId10002_part0_0_randomId10006",
    "piece": Object {
      "_id": "randomId10011",
      "adLibSourceId": "randomId10002_globalAdLib0",
      "disabled": false,
      "dynamicallyInserted": true,
      "enable": Object {
        "start": "now",
      },
      "externalId": "MOCK_GLOBAL_ADLIB_0",
      "infiniteId": "randomId10011",
      "infiniteMode": 3,
      "name": "Global AdLib 0",
      "outputLayerId": "pgm",
      "partId": "randomId10002_part0_0",
      "rundownId": "randomId10002",
      "sourceLayerId": "cam0",
      "status": -1,
      "timings": Object {
        "next": Array [],
        "playOffset": Array [],
        "startedPlayback": Array [],
        "stoppedPlayback": Array [],
        "take": Array [
          1000,
        ],
        "takeDone": Array [],
        "takeOut": Array [],
      },
    },
    "rundownId": "randomId10002",
  },
]
`;

exports[`Playout API Global and Part Ad-Libs 2`] = `
Array [
  Object {
    "_id": "randomId10002_part0_0_randomId10006_randomId10002_piece000",
    "partInstanceId": "randomId10002_part0_0_randomId10006",
    "piece": Object {
      "_id": "randomId10002_piece000",
      "enable": Object {
        "start": 0,
      },
      "externalId": "MOCK_PIECE_000",
      "name": "Piece 000",
      "outputLayerId": "pgm",
      "partId": "randomId10002_part0_0",
      "rundownId": "randomId10002",
      "sourceLayerId": "cam0",
      "status": 0,
    },
    "rundownId": "randomId10002",
  },
  Object {
    "_id": "randomId10002_part0_0_randomId10006_randomId10002_piece001",
    "partInstanceId": "randomId10002_part0_0_randomId10006",
    "piece": Object {
      "_id": "randomId10002_piece001",
      "enable": Object {
        "start": 0,
      },
      "externalId": "MOCK_PIECE_001",
      "name": "Piece 001",
      "outputLayerId": "pgm",
      "partId": "randomId10002_part0_0",
      "rundownId": "randomId10002",
      "sourceLayerId": "vt0",
      "status": 0,
    },
    "rundownId": "randomId10002",
  },
  Object {
    "_id": "randomId10002_part0_0_randomId10006_randomId10002_piece010",
    "partInstanceId": "randomId10002_part0_0_randomId10006",
    "piece": Object {
      "_id": "randomId10002_piece010",
      "enable": Object {
        "start": 0,
      },
      "externalId": "MOCK_PIECE_010",
      "name": "Piece 010",
      "outputLayerId": "pgm",
      "partId": "randomId10002_part0_0",
      "rundownId": "randomId10002",
      "sourceLayerId": "cam0",
      "status": 0,
    },
    "rundownId": "randomId10002",
  },
  Object {
    "_id": "randomId10002_part0_0_randomId10006_randomId10011",
    "partInstanceId": "randomId10002_part0_0_randomId10006",
    "piece": Object {
      "_id": "randomId10011",
      "adLibSourceId": "randomId10002_globalAdLib0",
      "disabled": false,
      "dynamicallyInserted": true,
      "enable": Object {
        "start": "now",
      },
      "externalId": "MOCK_GLOBAL_ADLIB_0",
      "infiniteId": "randomId10011",
      "infiniteMode": 0,
      "name": "Global AdLib 0",
      "originalInfiniteMode": 3,
      "outputLayerId": "pgm",
      "partId": "randomId10002_part0_0",
      "rundownId": "randomId10002",
      "sourceLayerId": "cam0",
      "status": -1,
      "timings": Object {
        "next": Array [],
        "playOffset": Array [],
        "startedPlayback": Array [],
        "stoppedPlayback": Array [],
        "take": Array [
          1000,
        ],
        "takeDone": Array [],
        "takeOut": Array [],
      },
      "userDuration": Object {
        "end": "#piece_group_randomId10013.start + 0",
      },
    },
    "rundownId": "randomId10002",
  },
  Object {
    "_id": "randomId10002_part0_0_randomId10006_randomId10013",
    "partInstanceId": "randomId10002_part0_0_randomId10006",
    "piece": Object {
      "_id": "randomId10013",
      "adLibSourceId": "randomId10002_adLib000",
      "disabled": false,
      "dynamicallyInserted": true,
      "enable": Object {
        "duration": 1000,
        "start": "now",
      },
      "externalId": "MOCK_ADLIB_000",
      "infiniteMode": 0,
      "name": "AdLib 0",
      "outputLayerId": "pgm",
      "partId": "randomId10002_part0_0",
      "rundownId": "randomId10002",
      "sourceLayerId": "vt0",
      "status": -1,
      "timings": Object {
        "next": Array [],
        "playOffset": Array [],
        "startedPlayback": Array [],
        "stoppedPlayback": Array [],
        "take": Array [
          3000,
        ],
        "takeDone": Array [],
        "takeOut": Array [],
      },
    },
    "rundownId": "randomId10002",
  },
]
`;
